sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/StringFormatValidatorBase.cs

```diff
@@ -0,0 +1,444 @@
+            var split = formatItem.Split(',', ':');
+
+            if (indexOfComma >= 0 && indexOfColon >= 0 && indexOfColon < indexOfComma ||
+                split.Length > 3)
+            {
+                return new ValidationFailure(ValidationFailure.FailureKind.FormatItemMalformed);
+            }
+
+            int index;
+            int? alignment = null;
+            string formatString = null;
+
+            if (!int.TryParse(split[0], out index))
+            {
+                return new ValidationFailure(ValidationFailure.FailureKind.FormatItemIndexIsNaN);
+            }
+
+            if (indexOfComma >= 0)
+            {
+                int localAlignment;
+                if (!int.TryParse(split[1], out localAlignment))
+                {
+                    return new ValidationFailure(ValidationFailure.FailureKind.FormatItemAlignmentIsNaN);
+                }
+                alignment = localAlignment;
+            }
+
+            if (indexOfColon >= 0)
+            {
+                formatString = indexOfComma >= 0
+                    ? split[2]
+                    : split[1];
+            }
+
+            formatStringItem = new FormatStringItem(index, alignment, formatString);
+            return null;
+        }
+
+        private static ValidationFailure TryValidateFormatString(string formatStringText,
+            ICollection<FormatStringItem> formatStringItems, ArgumentListSyntax argumentList, int formatArgumentIndex,
+            SemanticModel semanticModel)
+        {
+            var formatArguments = argumentList.Arguments
+                .Skip(formatArgumentIndex + 1)
+                .Select(arg => FormatStringArgument.Create(arg.Expression, semanticModel))
+                .ToList();
+            var maxFormatItemIndex = formatStringItems.Max(item => (int?)item.Index);
+
+            var realArgumentsCount = formatArguments.Count;
+            if (formatArguments.Count == 1 &&
+                formatArguments[0].TypeSymbol.Is(TypeKind.Array))
+            {
+                realArgumentsCount = formatArguments[0].ArraySize;
+                if (realArgumentsCount == -1)
+                {
+                    // can't statically check the override that supplies args in an array variable
+                    return null;
+                }
+            }
+
+            return IsFormatValidSafetyNet(formatStringText) ??
+                IsSimpleString(formatStringItems.Count, realArgumentsCount) ??
+                HasFormatItemIndexTooBig(maxFormatItemIndex, realArgumentsCount) ??
+                HasMissingFormatItemIndex(formatStringItems, maxFormatItemIndex) ??
+                HasUnusedArguments(formatArguments, maxFormatItemIndex);
+        }
+
+
+        private static ValidationFailure IsFormatValidSafetyNet(string formatString)
+        {
+            try
+            {
+                var _ = string.Format(formatString, new object[1000000]);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Remove this unused '_' local variable. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1481)